### ISSUE TITLE
Tests use a lot of duplicate points

### ISSUE THREAD

**Issue by betatim** [scikit-optimize/scikit-optimize#302]
Since we added the warning about re-evaluating the objective at the same point our tests are full of warnings:
```
scikit-optimize/skopt/optimizer/optimizer.py:174: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated at this point before.")
```

We should think whether we want t mute the warning or improve the tests.

**Comment by glouppe** [comment]
or we might check why some points are reevaluated.

**Comment by Henri-Lo** [comment]
How can you suppress the warning? I am running the very first example on the doc and I am getting the warning, perhaps you could update the example in the doc with suppressed warning?

UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "

https://github.com/scikit-optimize/scikit-optimize

**Comment by TropComplique** [comment]
The same is here. A lot of annoying warnings. What to do with them?

**Comment by TropComplique** [comment]
@Henri-Lo So, you can suppress the warnings by using:
```
import warnings
warnings.filterwarnings('ignore', message='The objective has been evaluated at this point before.')
```
But is it the right thing to do?

**Comment by MechCoder** [comment]
The right thing to do for the example, tests or in general?

**Comment by TropComplique** [comment]
Hi. 

In general. When using `skopt` for hyperparameter optimization, for example.  
Are there warnings because I'm using `skopt` in a wrong way or because `skopt` is not fully developed yet?

**Comment by MechCoder** [comment]
No package is fully-developed ever :) Can you post a minimal code-snippet?

It is not too uncommon in-case you are optimizing over a space that is completely categorical for instance.

**Comment by betatim** [comment]
We should try and remove the warnings from the tests by either choosing bigger spaces or filtering them where we know that it isn't a problem.

In general the warning tells you that "there might be a problem". Usually evaluating the same point multiple times isn't something you want to do, but if you have a (very) noisy objective it makes sense to do so. In my experience I usually get the warning when I get the type of one of the dimensions wrong, for example `(2, 2.1443)` as a dimension gives you an `Integer` dimension (-> #391) when really I wanted a floating point dimension.

**Comment by AzraelDD** [comment]
Hi everybody,

is this still being considered? There should be (if this is possible) some sort of input flag for ``optimizer.ask()`` that forces it to avoid points that have been evaluated before. Instead it should evaluate the next "best" / "most interesting" points or maybe even some totally random points.

In my use case I do optimization in parallel just like in this example: https://scikit-optimize.github.io/notebooks/parallel-optimization.html

My function to evalute the points involves a simulation that takes several minutes each, so running the simulations in parallel is a huge win for me. Most of the time I am not searching for the single "best" solution of my evaluation function, but for e.g. "all solutions of my evaluation function with an error less than 1%". In an optimization problem with two dimensions this gives me a whole "solution set". For example, with dimensions x and y, the function might be true for increasing x while y decreases.

To get the broad spectrum of solutions I need to keep the optimization loop running until I (manually) decide that most of the solution "space" has been explored. After many hundreds of evaluations, the ``optimizer.ask()`` sometimes seems to get stuck near an assumed global minima and just keeps repeating the evaluation of points there.

Consider the example output of ``skopt.plots.plot_objective()`` below. What I need as a result is the yellow area. Each point in there can be a valid solution to my evaluation function. But the optimizer seems to get stuck near the bottom right. This example is very well explored. But sometimes the optimizer gets stuck to early. Starting with a larger value of ``n_initial_points`` helps. But a flag that specifically makes the optimizer not evaluate any points twice would help, too. 

![skopt_objective](https://user-images.githubusercontent.com/8526498/48930784-0e595800-eef3-11e8-8df7-8dbff35e216b.png)

**Comment by rpcoelho17** [comment]
I agree with the above post that points that have already been evaluated should be skiped.  I work with neural networks that take 45 minutes to evaluate 1 single time.  If the work has been done already for a certain point it should not call the evaluation function for that point but simply move to the next best point.  Doesn't hat make sense?

**Comment by bionicles** [comment]
yes this makes sense and should be a key feature to not repeat trials unless we've already tried everything and just want to reduce variance

**Comment by owenljn** [comment]
Hmmmmm, 2 years have passed, looks like this is still an issue.

**Comment by saniaki** [comment]
Still hasn't fixed? I am having the same issue. Try the same points multiple times which is not an efficient way of doing optimization.

**Comment by BabaYara** [comment]
I think the best solution at this point is to write a skip if statement in your function and just pass it the optimizer the result from the previous run if it wants to re-run a certain point.

**Comment by kernc** [comment]
Another workaround is to use [`DeltaXStopper`](https://scikit-optimize.github.io/stable/modules/generated/skopt.callbacks.DeltaXStopper.html#skopt.callbacks.DeltaXStopper) callback:
```py
gp_minimize(..., callback=DeltaXStopper(1e-8))
```

**Comment by jnettels** [comment]
Hello everyone, I proposed a workaround to this issue in PR #1050:

> #302 discusses a long standing issue where the ask() function to "Query point or multiple points at which objective should be evaluated" will just repeatedly return the same point(s) to evaluate. In cases where the objective always returns the same results, the solver can even get stuck in a loop.
> 
> This PR aims to mitigate the issue. Instead of just warning the user about evaluating a point twice, use a new random point instead (and include this information in the warning).
> 
> This is not a good solution. Ideally, the next best point should be evaluated instead, but I do not know how to do that.
> 
> Also, in #302 a user said "Usually evaluating the same point multiple times isn't something you want to do, but if you have a (very) noisy objective it makes sense to do so." So the solution presented here might actually cause a problem for users who want the same point to be evaluated multiple times.
> 
> Maybe this just stays here as a reference for users who want a very easy workaround that they can implement themselves.

Some possible workarounds were mentioned in this thread before, but my PR presents one that every user can apply on their own if they need to. Maybe this inspires someone to create a real fix for this issue...

**Comment by kart-codes** [comment]
I have a solution , it kind of works. It does not repeat any runs. But sometimes may not run all parallel jobs.
```
from skopt import Optimizer
from skopt.space import Real, Integer, Space
from joblib import Parallel, delayed
from skopt.benchmarks import branin
import skopt
import warnings

#set optimizer properties
optimizer=Optimizer(
    dimensions=[Real(0.34,0.75),Real(0.36,0.78)],
    random_state=1,
    base_estimator='gp')
warnings.filterwarnings('ignore', message='The objective has been evaluated at this point before.')
#defined test function
def branined(x_ins):
    return branin(x_ins[1:])
#Set runs
last_count=0
y_lookup=[]
for i in range(10):
    x=optimizer.ask(n_points=4) #x is a list of n_points points
    ### group non-repeat x as x_runs and repeats as x_non
    x_run=[val for val in x if val not in optimizer.Xi]
    x_non=[val for val in x if val in optimizer.Xi]
    #run only for non repeat x's that is x_run
    if len(x_run)!=0:
        #job_counter to keep track of 
        job_counter=list(range(1+last_count,len(x_run)+last_count+1))
        x_added=[[job_counter[x_ind]]+x_val for x_ind,x_val in enumerate(x_run)]
        y_run=Parallel(n_jobs=len(x_run))(delayed(branined)(v) for v in x_added) #evaluate points in parallel
        #store for lookup, to identify job_count corresponding to minimum yi
        y_lookup.extend([[job_counter[y_ind],y_val] for y_ind,y_val in enumerate(y_run)])
        last_count+=len(x_non)
    else:
        y_run=[]

    y_non=[optimizer.yi[optimizer.Xi.index(x_val)] for x_val in x_non]
    x_update=x_run+x_non
    y_update=y_run+y_non
    ###
    optimizer.tell(x_update,y_update)
	
#Optimized x & Y values
y_loc=optimizer.yi.index(min(optimizer.yi))
print(min(optimizer.yi))
print(optimizer.Xi[y_loc])

```

**Comment by Jose-Verdu-Diaz** [comment]
Is there any update on this issue?

**Comment by fingoldo** [comment]
> ```python
> callback=DeltaXStopper(1e-8)
> ```

while this gives less of those annoying warnings, results in my testing became worse (true extremum found less frequently).
